main.c

```diff
@@ -331,12 +380,75 @@ multithreaded(struct task_t *task, struct config_t *config)
     return context.found;
 }
 
+void *
+gn_worker(void *arg)
+{
+    struct gn_context_t *context = (struct gn_context_t *) arg;
+    struct config_t *config = context->config;
+
+    struct st_context_t st_context;
+    st_context.hash = context->hash;
+    st_context.cd.initialized = 0;
+
+    while (true)
+    {
+        struct task_t task;
+        pthread_mutex_lock(&context->mutex);
+        task = *context->iter_state.task;
+        bool has_next = iter_next(&context->iter_state);
+        pthread_mutex_unlock(&context->mutex);
+
+        bool found = process_task(&task, config, &st_context);
+        if (found)
+        {
+            memcpy(context->password, task.password, sizeof(task.password));
+            context->found = true;
+        }
+
+        if (!has_next || context->found) break;
+    }
+    return NULL;
+}
+
+bool
+generator(struct task_t *task, struct config_t *config)
+{
+    struct gn_context_t context;
+    context.hash = config->hash;
+
+    task->from = 2;
+    task->to = config->length;
+    iter_init(&context.iter_state, task, config->alphabet);
+
+    pthread_mutex_init(&context.mutex, NULL);
+    context.password[0] = 0;
+    context.config = config;
+
+    int cpu_count = sysconf(_SC_NPROCESSORS_ONLN) - 1;
+    pthread_t threads[cpu_count];
+    for (int i = 0; i < cpu_count; ++i)
```

Comment:
It make sense to start from 1, just because we will have one more worker in main thread.